apps/web/app/(app)/mail/page.tsx

```diff
@@ -20,65 +20,80 @@ export default function Mail({
     ? { type: searchParams.type }
     : {};
 
-  const { data, isLoading, error, mutate } = useSWR<ThreadsResponse>(
-    `/api/google/threads?${new URLSearchParams(query as any).toString()}`,
-    {
+  const getKey = (
+    pageIndex: number,
+    previousPageData: ThreadsResponse | null,
+  ) => {
+    if (previousPageData && !previousPageData.nextPageToken) return null;
+    // For the first page, use the base query
+    if (pageIndex === 0)
+      return `/api/google/threads?${new URLSearchParams(query as Record<string, string>).toString()}`;
+    // For subsequent pages, append the nextPageToken
+    return `/api/google/threads?nextPageToken=${previousPageData?.nextPageToken}&${new URLSearchParams(query as any).toString()}`;
+  };
+
+  const { data, size, setSize, isLoading, error, mutate } =
+    useSWRInfinite<ThreadsResponse>(getKey, {
       keepPreviousData: true,
       dedupingInterval: 1_000,
-    },
-  );
+      revalidateOnFocus: false,
+    });
 
+  const allThreads = data ? data.flatMap((page) => page.threads) : [];
+  const isLoadingMore =
+    isLoading || (size > 0 && data && typeof data[size - 1] === "undefined");
+  const showLoadMore = data ? !!data[data.length - 1]?.nextPageToken : false;
+
+  // store `refetch` in the atom so we can refresh the list upon archive via command k
+  // TODO is this the best way to do this?
   const refetch = useCallback(
     (removedThreadIds?: string[]) => {
-      mutate(undefined, {
-        rollbackOnError: true,
-        optimisticData: (currentData) => {
-          if (!removedThreadIds)
-            return {
-              threads: currentData?.threads || [],
-              nextPageToken: undefined,
-            };
-          const threads =
-            currentData?.threads.filter(
+      mutate(
+        (currentData) => {
+          if (!currentData) return currentData;
+          if (!removedThreadIds) return currentData;
+
+          return currentData.map((page) => ({
+            ...page,
+            threads: page.threads.filter(
               (t) => !removedThreadIds.includes(t.id),
-            ) || [];
-          return { threads, nextPageToken: undefined };
+            ),
+          }));
         },
```

Comment:
Hey, I realise this was messy code before. But what's the reason the optimistic update was removed?

Reply:
But also if it works from your video, then maybe it's fine.

Reply:
Just for the simplicity of the logic and code.